src/Dictionaries/SSDCacheDictionary.cpp

```diff
@@ -0,0 +1,1698 @@
+    }
+
+    DECLARE(UInt8)
+    DECLARE(UInt16)
+    DECLARE(UInt32)
+    DECLARE(UInt64)
+    DECLARE(UInt128)
+    DECLARE(Int8)
+    DECLARE(Int16)
+    DECLARE(Int32)
+    DECLARE(Int64)
+    DECLARE(Float32)
+    DECLARE(Float64)
+    DECLARE(Decimal32)
+    DECLARE(Decimal64)
+    DECLARE(Decimal128)
+#undef DECLARE
+
+#define DECLARE(TYPE) \
+    void SSDCacheDictionary::get##TYPE( \
+        const std::string & attribute_name, \
+        const PaddedPODArray<Key> & ids, \
+        const PaddedPODArray<TYPE> & def, \
+        ResultArrayType<TYPE> & out) const \
+    { \
+        const auto index = getAttributeIndex(attribute_name); \
+        checkAttributeType(name, attribute_name, dict_struct.attributes[index].underlying_type, AttributeUnderlyingType::ut##TYPE); \
+        getItemsNumberImpl<TYPE, TYPE>( \
+            index, \
+            ids, \
+            out, \
+            [&](const size_t row) { return def[row]; }); \
+    }
+    DECLARE(UInt8)
+    DECLARE(UInt16)
+    DECLARE(UInt32)
+    DECLARE(UInt64)
+    DECLARE(UInt128)
+    DECLARE(Int8)
+    DECLARE(Int16)
+    DECLARE(Int32)
+    DECLARE(Int64)
+    DECLARE(Float32)
+    DECLARE(Float64)
+    DECLARE(Decimal32)
+    DECLARE(Decimal64)
+    DECLARE(Decimal128)
+#undef DECLARE
+
+#define DECLARE(TYPE) \
+    void SSDCacheDictionary::get##TYPE( \
+        const std::string & attribute_name, \
+        const PaddedPODArray<Key> & ids, \
+        const TYPE def, \
+        ResultArrayType<TYPE> & out) const \
+    { \
+        const auto index = getAttributeIndex(attribute_name); \
+        checkAttributeType(name, attribute_name, dict_struct.attributes[index].underlying_type, AttributeUnderlyingType::ut##TYPE); \
+        getItemsNumberImpl<TYPE, TYPE>( \
+            index, \
+            ids, \
+            out, \
+            [&](const size_t) { return def; }); \
+    }
+    DECLARE(UInt8)
+    DECLARE(UInt16)
+    DECLARE(UInt32)
+    DECLARE(UInt64)
+    DECLARE(UInt128)
+    DECLARE(Int8)
+    DECLARE(Int16)
+    DECLARE(Int32)
+    DECLARE(Int64)
+    DECLARE(Float32)
+    DECLARE(Float64)
+    DECLARE(Decimal32)
+    DECLARE(Decimal64)
+    DECLARE(Decimal128)
+#undef DECLARE
+
+template <typename AttributeType, typename OutputType, typename DefaultGetter>
+void SSDCacheDictionary::getItemsNumberImpl(
+        const size_t attribute_index, const PaddedPODArray<Key> & ids, ResultArrayType<OutputType> & out, DefaultGetter && get_default) const
+{
+    const auto now = std::chrono::system_clock::now();
+
+    std::unordered_map<Key, std::vector<size_t>> not_found_ids;
+    storage.getValue<OutputType>(attribute_index, ids, out, not_found_ids, get_default, now);
+    if (not_found_ids.empty())
+        return;
+
+    std::vector<Key> required_ids(not_found_ids.size());
+    std::transform(std::begin(not_found_ids), std::end(not_found_ids), std::begin(required_ids), [](const auto & pair) { return pair.first; });
+
+    storage.update(
```

Comment:
Maybe better to wrap it in a separate function, which adds source_ptr and lifetime and call this function?